### DISCUSSION TITLE
How to use the redirect_uri parameter in github apps with multiple callback urls

### DISCUSSION THREAD

**Discussion by leecalcote** [community/community#54273] [Apps, API and Webhooks] [answered]
### Select Topic Area

Question

### Body

When developing GitHub Apps, is it common practice to create one app per deployment environment (e.g. dev, staging, production)? 

If not, are there examples of how the [redirect_uri parameter](https://docs.github.com/en/apps/creating-github-apps/setting-up-a-github-app/about-the-user-authorization-callback-url) is used to route authentication sequences when multiple callback urls are used in the GitHub App?

**Comment by akatyabg** [comment, marked as answer]
I don't know about the common best practices but this is how you can use multiple redirect URLs for authenticating users: https://docs.github.com/en/apps/creating-github-apps/guides/building-a-login-with-github-button-with-a-github-app#handle-multiple-callback-urls

Example: In your Github App configurations define a callback url for each envirnment:
<img width="794" alt="image" src="https://user-images.githubusercontent.com/6072782/235860833-e81d60b7-7b98-484d-b432-73c95ba6f4a2.png">
Then when directing the user, add the wanted redirect_url as a query paramenter e.g. `https://github.com/login/oauth/authorize?client_id=12345&state=abcdefg&redirect_url=http%3A%2F%2Fprod.environment.com%2Fcallback`. This will then redirect the user to the production envionment.

As described in the documentation if you don't specify a `redirect_url ` or an exact match is not found, the first configured url will be used. In the example it would be` http://dev.envirnoment.com/callback`

Hope this helps.

**Reply by leecalcote** [reply]
Much appreciated, @akatyabg 👍

**Reply by merijn-weiss** [reply]
Please note that this example provides a typo (twice). The parameter is **redirect_uri** and not **redirect_url**.

**Reply by Hexi1997** [reply]
redirect_uri does not work when i click cancel

**Reply by Yashasv-Prajapati** [reply]
Does this also work if the user has already installed my app? In my case this does not work and it rather brings me to the configure page of the app.

**Reply by rtpa25** [reply]
although this does not work when i use the install url ex: `${env.NEXT_PUBLIC_GITHUB_IMPORTER_APP_PUBLIC_LINK}/installations/select_target?redirect_uri=${redirectUrl}`
